pkg/controller/factory/base_controller.go

```diff
@@ -36,36 +38,54 @@ func (c *baseController) Run(ctx context.Context, workers int) {
 		panic("timeout waiting for informer cache") // this will be recovered using HandleCrash()
 	}
 
-	var workerWaitGroup sync.WaitGroup
+	var workerWg sync.WaitGroup
 	defer func() {
 		defer klog.Infof("All %s workers have been terminated", c.name)
-		workerWaitGroup.Wait()
+		workerWg.Wait()
 	}()
 
 	// queueContext is used to track and initiate queue shutdown
 	queueContext, queueContextCancel := context.WithCancel(context.TODO())
 
 	for i := 1; i <= workers; i++ {
 		klog.Infof("Starting #%d worker of %s controller ...", i, c.name)
-		workerWaitGroup.Add(1)
+		workerWg.Add(1)
 		go func() {
 			defer func() {
 				klog.Infof("Shutting down worker of %s controller ...", c.name)
-				workerWaitGroup.Done()
+				workerWg.Done()
 			}()
 			c.runWorker(queueContext)
 		}()
 	}
 
 	// runPeriodicalResync is independent from queue
 	if c.resyncEvery > 0 {
-		workerWaitGroup.Add(1)
+		workerWg.Add(1)
 		go func() {
-			defer workerWaitGroup.Done()
+			defer workerWg.Done()
 			c.runPeriodicalResync(ctx, c.resyncEvery)
 		}()
 	}
 
+	// run post-run hooks (custom triggers, etc.)
+	if len(c.postRunHooks) > 0 {
+		var hookWg sync.WaitGroup
+		defer func() {
+			defer klog.Infof("All %s post-run hooks have been terminated", c.name)
+			hookWg.Wait() // wait for the post-run hooks
+		}()
+		for i := range c.postRunHooks {
+			hookWg.Add(1)
+			go func(index int) {
+				defer hookWg.Done()
+				if err := c.postRunHooks[index](ctx, c.syncContext); err != nil {
```

Comment:
are you sure you want to call them here and not after the context closed?!

Reply:
@sttts we have two controllers that are adding "custom" triggers (if A change add to workqueue...). Maybe "PostRun" is not the perfect name for this... This hook should run after the controller is started... "PostStart" ?

Reply:
I think I got it. The name postRun is super bad. This is not post run. It is post start. Please rename it to avoid confusion.

Reply:
renamed.